build-tools/Xamarin.Android.Tools.BootstrapTasks/Xamarin.Android.Tools.BootstrapTasks/GenerateSupportedPlatforms.cs

```diff
@@ -0,0 +1,70 @@
+using System.IO;
+using System.Linq;
+using System.Xml;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+namespace Xamarin.Android.Tools.BootstrapTasks
+{
+	/// <summary>
+	/// Generates Microsoft.Android.Sdk.SupportedPlatforms.props
+	/// Similar to: https://github.com/dotnet/sdk/blob/18ee4eac8b3abe6d554d2e0c39d8952da0f23ce5/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.WindowsSupportedTargetPlatforms.props
+	/// </summary>
+	public class GenerateSupportedPlatforms : Task
+	{
+		/// <summary>
+		/// A list of AndroidApiInfo.xml files produced by Mono.Android.targets
+		/// </summary>
+		[Required]
+		public string [] AndroidApiInfo { get; set; }
+
+		/// <summary>
+		/// The output file to generate
+		/// </summary>
+		[Required]
+		public string OutputFile { get; set; }
+
+		public override bool Execute ()
+		{
+			var versions = new AndroidVersions (
+				AndroidApiInfo.Select (d => Path.GetDirectoryName (d)));
+			var settings = new XmlWriterSettings {
+				OmitXmlDeclaration = true,
+				Indent = true,
+			};
+			using (var writer = XmlWriter.Create (OutputFile, settings)) {
+
+				writer.WriteComment (@"
+***********************************************************************************************
+Microsoft.Android.Sdk.SupportedPlatforms.props
+
+Specifies the supported Android platform versions for this SDK.
+
+***********************************************************************************************
+");
+				writer.WriteStartElement ("Project");
+
+				writer.WriteStartElement ("PropertyGroup");
+				writer.WriteStartElement ("TargetPlatformVersion");
+				writer.WriteAttributeString ("Condition", " '$(TargetPlatformVersion)' == '' ");
+				writer.WriteString (versions.MaxStableVersion.ApiLevel.ToString ());
+				writer.WriteEndElement (); // </TargetPlatformVersion>
+				writer.WriteEndElement (); // </PropertyGroup>
+
+				writer.WriteStartElement ("ItemGroup");
+				foreach (AndroidVersion version in versions.InstalledBindingVersions) {
+					writer.WriteStartElement ("AndroidSupportedTargetPlatform");
+					writer.WriteAttributeString ("Include", version.ApiLevel.ToString ());
+					writer.WriteEndElement (); // </AndroidSupportedTargetPlatform>
+				}
+				writer.WriteStartElement ("SupportedTargetPlatform");
+				writer.WriteAttributeString ("Condition", " '$(TargetPlatformIdentifier)' == 'Android' ");
+				writer.WriteAttributeString ("Include", "@(AndroidSupportedTargetPlatform)");
+
+				writer.WriteEndDocument (); // </Project>
```

Comment:
Once we generate these item groups, we get:

https://github.com/dotnet/sdk/blob/a0e17b2b5d2ec75c16e1083b1f926a258126c73a/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.BeforeCommon.targets#L188-L198

This fills out `$(DefineConstants)` with `ANDROID;ANDROID21;...;ANDROID30`.

@dellis1972 @jonpryor do we need to get rid of the old `__ANDROID__` defines? The .NET 5 design spec doesn't list these:

https://github.com/dotnet/designs/blob/master/accepted/2020/net5/net5.md#preprocessor-symbols

So should we stop running this task?

https://github.com/xamarin/xamarin-android/blob/e61ff790ca524ac173be3965fbbad510d8303b43/src/Xamarin.Android.Build.Tasks/Tasks/GetAndroidDefineConstants.cs#L34-L38

Reply:
It would make it easier to port code if you left it in, but then there would be a bunch of duplicate constants.

Reply:
So we have a few options. 

1. a hard break , users need to switch to the new defines for net5. (not desirable I think).
2. define the old `__` prefixed ones in net5..
3. define the new non prefixed ones in our product now as well as the old, but issue warnings the old are deprecated (if possible) to allow users to migrate. We can then phase our the old style.

Personally I would prefer option 3.

Reply:
I don't see a way to deprecate usage of `#if __ANDROID__`, etc.?

It would have to be something in Roslyn, because  it would require looking through every C# source file.

I guess we leave both old & new `$(DefineConstants)` for now.

Reply:
It occurs to me that Roslyn is now shipping a standard set of Analyzers in the .NET SDK, so if we wanted to we could probably add an analyzer there that detected usage of the old Android compilation constants format.

Reply:
@jonathanpeppers [asked](https://github.com/xamarin/xamarin-android/pull/5007#discussion_r470265770)

> do we need to get rid of the old __ANDROID__ defines?

***No***, let's not. :-)